### PR TITLE
Postcode lookup

### PR DESCRIPTION
**Pull request by neil-stephen-mcgonigle**
- add postcode lookup client-side progressive enhancement elements
- add cica web proxied calls to the os-places api to retrieve addresses
- add new secret for os-places api key

### CODE REVIEW

File: src/modules/postcode-lookup/test/postcode-lookup.test.jsdom.js [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+                postcodeLookup = createPostcodeLookup(window);
+                await postcodeLookup.init();
+                expect(
+                    window.document.getElementById('fill-out-the-fields-manually-hint')
+                ).toBeNull();
+            });
+        });
+        describe('on postcode lookup find address button click', () => {
+            describe('200', () => {
+                it('Should return 3 results for valid postcode', async () => {
```
[the remark this answers is not in the document]
Comment (neil-stephen-mcgonigle) [reply src/modules/postcode-lookup/test/postcode-lookup.test.jsdom.js:158]: I have refactored the tests.

File: address-finder/address-finder-service.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+
+const service = require('../questionnaire/request-service')();
```
Comment (armoj) [inline address-finder/address-finder-service.js:3]: Thoughts on changing this to:
`const {get} = require('../questionnaire/request-service')();`

File: kube_deploy/Dev/deploy.yml [hunk 3]
```diff
@@ -74,6 +74,11 @@ spec:
                 secretKeyRef:
                   name: cica-web-secrets
                   key: cw_maintenance_message_enabled
+            - name: CW_OS_PLACES_API_KEY
```
Comment (armoj) [inline kube_deploy/Dev/deploy.yml:77]: Given this is visible in the request header, does this need to be a bona fide secret?

File: src/modules/postcode-lookup/index.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+ * Returns a postocde lookup module.
+ * @param {Object} window  Global window object
+ */
+function createPostcodeLookup(window) {
+    // Define an empty JSON object to temporarily store matched address results.
+    // eslint-disable-next-line no-unused-vars
+    let tmpAddressSearchResultsJson = {};
+
+    function addSearchResultsToSelectElement(data) {
+        const selectElementResults = window.document.getElementById(
```
Comment (armoj) [inline src/modules/postcode-lookup/index.js:11]: cache this outside of the `addSearchResultsToSelectElement` function. Saves querying the DOM each time the function is called.

File: src/modules/postcode-lookup/index.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+            'address-search-results-dropdown'
+        );
+        // Clear the results drop-down list.
+        const options = window.document.querySelectorAll('#address-search-results-dropdown option');
```
Comment (armoj) [inline src/modules/postcode-lookup/index.js:15]: [getElementsByTagName](https://developer.mozilla.org/en-US/docs/Web/API/Element/getElementsByTagName) returns a Live HTMLCollection so you could do something like:

`selectElementResults.getElementsByTagName('options');` and this list would be updated on the fly automatically. So you could move this to outside this function also and you would not need to activelly query the DOM again.

File: src/modules/postcode-lookup/index.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        selectElementResults.add(option);
+
+        // See also {@link https://apidocs.os.uk/reference/os-places-dpa-output}
+        const dataset = 'DPA';
+        // Loop through the data results; adding the address string as a new option
+        // to the results drop-down list.
+        data.results.forEach(function addElemets(val, i) {
+            const opt = window.document.createElement('option');
+            opt.value = i;
+            opt.text = val[dataset.toUpperCase()].ADDRESS;
```
Comment (armoj) [inline src/modules/postcode-lookup/index.js:34]: redundant call to `.toUpperCase()` here

File: src/modules/postcode-lookup/index.js [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            )
+        );
+        descriptionDiv.setAttribute('id', 'fill-out-the-fields-manually-hint');
+        descriptionDiv.setAttribute('class', 'govuk-hint');
+        window.document
+            .querySelector('.govuk-form-group')
+            .insertAdjacentElement('beforeBegin', descriptionDiv);
+    }
+
+    function createPostcodeSearchElements() {
```
Comment (armoj) [inline src/modules/postcode-lookup/index.js:80]: Would it be easier to just have a chunk of HTML that you append? i.e:

const postcodeDiv = window.document.createElement('div');
postcodeDiv.innerHTML = '<div id="postcode-search" class="govuk-form-group">..All other html in string literal form..</div>';
window.document
            .getElementById('fill-out-the-fields-manually-hint').insertAdjacentElement(postcodeDiv);

File: kube_deploy/Dev/deploy.yml (hunk 3 above)
Comment (neil-stephen-mcgonigle) [reply kube_deploy/Dev/deploy.yml:77]: Yeah, requests the os-api are proxied via cica-web and will be sent over https. And we don't want to expose this key in our repo.

File: src/modules/postcode-lookup/index.js (hunk 7 above)
Comment (neil-stephen-mcgonigle) [reply src/modules/postcode-lookup/index.js:80]: See the Security considerations [here](https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML)

**Pull request closed without merging**